lib/huginn_acumen_product_agent/concerns/acumen_product_query_concern.rb

```diff
@@ -71,6 +71,28 @@ def get_product_contributors(acumen_client, products)
         products
     end
 
+    def get_master_products_by_id(client, products, physical_formats, digital_formats)
+        # for each product check if wrapper product is master product and active
+        # if false store the id of the active master product
+        # run it through get_products_by_ids and add it to products
+        products.each do |product|
+            wrapper_id = product['identifier']
+            correct_id = 0
+            product['model'].each do |variant|
+                if variant['acumenAttributes']['is_master'] && variant['isAvailableForPurchase']
+                    correct_id = variant['identifier']
+                end
+            end
+            if wrapper_id != correct_id && correct_id != 0
+                products.delete(product)
+                products.push(get_products_by_ids(client, [correct_id.to_s])[0])
+                products = get_product_variants(client, products, physical_formats, digital_formats)
+            end
+        end
+
+        products
+    end
```

Comment:
This needs to be reworked.

I see what you're trying to accomplish, and I think the goal is correct, but you're calling `delete()` inside a loop -- effectively manipulating the contents of the `products` array while you're iterating over it. That's an accident waiting to happen.

The second issue here is that you're generating a lot of excess work because each time you "reload" a product, you are passing the _full list_ of products into `get_product_variants` -- meaning that you are refetching and reprocessing _every variant_ for _every product_ each time this block executes. You shouldn't have to call that variant fetch at all -- because you've already done that.

The `model` set for `ABCE` with variant `ABCP` will be the same as the `model` set for `ABCP` with a vairant of `ABCE`.

Something like the following is probably closer to what we want here (but it will need to be tested):

```
def get_master_products_by_id(client, products, physical_formats, digital_formats)
    #  Ensure that the _master_ product returned from this agent is the same regardless of
    #  which variant ID is used to fetch the collection of formats.
    #
    #  NOTE: Unlike the base_sku, the master product can change over time and is used as
    #  the authoritative source of core product attributes.

    master_products = []
  
    #  For each product, verify wether the root data represents the master variant, and if not,
    #  reload that product with the correct root data.
    products.each do |product|
      wrapper_id = product['identifier']
      master_id = 0
      product['model'].each do |variant|
        if variant['acumenAttributes']['is_master'] && variant['isAvailableForPurchase']
          correct_id = variant['identifier']
        end
      end

      if (wrapper_id == master_id || master_id == 0)
        #  The root `product` represents the master `product['model']`
        master_products.push(product)
      else
        #  The root `product` represents a variant `product['model']`

        #  Check to see if the intended product is already in the `master_products` array
        #
        #  NOTE: Conventionally it shouldn't be, but this avoids extra requests
        #  caused by user error when creating/modifying products in Acumen
        if (master_products.find { |p| p['identifier'] == remaster_id }).nil?
          reloaded_product = get_products_by_ids(client, [master_id.to_s])[0]

          #  If we found a product, add it to `master_products`
          unless reloaded_product.nil?
            #  `product['model']` already contains all the variants for this product set.
            #  update the reloaded product's `model` array.
            reloaded_product['model'] = product['model']
            master_products.push(product)
          end
        end
      end
    end

    master_products
end
```